sensor/common/detector/baseline/baseline.go

```diff
@@ -88,3 +102,170 @@ func (w *baselineEvaluator) IsOutsideLockedBaseline(pi *storage.ProcessIndicator
+	deploymentBaselines map[string]map[string]string
+	// content hash -> reference count and StringSet for deduplication
+	processSets map[string]*processSetEntry
+	// lock for thread safety
+	lock sync.RWMutex
+}
+
+type processSetEntry struct {
+	refCount  int
+	processes set.StringSet
+}
+
+// newOptimizedBaselineEvaluator creates the optimized baseline evaluator implementation
+func newOptimizedBaselineEvaluator() Evaluator {
+	return &optimizedBaselineEvaluator{
+		deploymentBaselines: make(map[string]map[string]string),
+		processSets:         make(map[string]*processSetEntry),
+	}
+}
+
+// removeReference decrements reference count and cleans up if necessary
+func (oe *optimizedBaselineEvaluator) removeReference(contentHash string) {
+	entry := oe.processSets[contentHash]
+	if entry == nil {
+		return // Entry doesn't exist
+	}
+
+	// Decrement reference count
+	entry.refCount--
+
+	// Clean up if no longer referenced
+	if entry.refCount <= 0 {
+		delete(oe.processSets, contentHash)
+	}
+}
+
+// computeProcessSetHash creates a deterministic hash for a process set
+func computeProcessSetHash(processes set.StringSet) string {
+	// Convert to sorted slice for deterministic hashing
+	processSlice := processes.AsSlice()
+	slices.Sort(processSlice)
+
+	// Create hash of concatenated processes
+	content := strings.Join(processSlice, "\n")
+	hash := sha256.Sum256([]byte(content))
+	return fmt.Sprintf("%x", hash)
+}
+
+// RemoveDeployment removes a deployment from the optimized baseline evaluator
+func (oe *optimizedBaselineEvaluator) RemoveDeployment(id string) {
+	oe.lock.Lock()
+	defer oe.lock.Unlock()
+
+	containerMap := oe.deploymentBaselines[id]
+	if containerMap == nil {
+		return
+	}
+
+	// Decrement reference counts for all process sets used by this deployment
+	for _, contentHash := range containerMap {
+		oe.removeReference(contentHash)
+	}
+
+	delete(oe.deploymentBaselines, id)
+	log.Debugf("Successfully removed deployment %s from baseline", id)
+}
+
+// AddBaseline adds a process baseline to the optimized evaluator
+func (oe *optimizedBaselineEvaluator) AddBaseline(baseline *storage.ProcessBaseline) {
+	oe.lock.Lock()
+	defer oe.lock.Unlock()
+
+	deploymentID := baseline.GetKey().GetDeploymentId()
+	containerName := baseline.GetKey().GetContainerName()
+
+	// Check if baseline is locked (has UserLockedTimestamp)
+	if baseline.GetUserLockedTimestamp() == nil {
+		// Unlocked baseline - remove any existing baseline (processes are always considered within unlocked baselines)
+		if oe.deploymentBaselines[deploymentID] != nil {
+			if oldContentHash, exists := oe.deploymentBaselines[deploymentID][containerName]; exists {
+				oe.removeReference(oldContentHash)
+				delete(oe.deploymentBaselines[deploymentID], containerName)
+				if len(oe.deploymentBaselines[deploymentID]) == 0 {
+					delete(oe.deploymentBaselines, deploymentID)
+				}
+			}
+		}
+		log.Debugf("Successfully added unlocked process baseline %s", baseline.GetId())
+		return
```

Comment:
Maybe a good candidate to extract to a separate func? sth like `handleLockedBaseline`

Reply:
pulled this out into removeReference